curvefs/src/client/fuse_client.cpp

```diff
@@ -561,5 +594,140 @@ CURVEFS_ERROR FuseClient::FuseOpRelease(fuse_req_t req, fuse_ino_t ino,
+CURVEFS_ERROR RenameOperator::CheckOverwrite() {
+    if (dstDentry_.flag() & DentryFlag::TYPE_FILE_FLAG) {
+        return CURVEFS_ERROR::OK;
+    }
+
+    std::list<Dentry> dentrys;
+    auto rc = dentryManager_->ListDentry(dstDentry_.inodeid(), &dentrys, 1);
+    if (rc == CURVEFS_ERROR::OK && !dentrys.empty()) {
+        LOG(ERROR) << "The directory is not empty"
+                   << ", dentry = (" << dstDentry_.ShortDebugString() << ")";
+        rc = CURVEFS_ERROR::NOTEMPTY;
+    }
+
+    return rc;
+}
+
+CURVEFS_ERROR RenameOperator::Precheck() {
+    auto rc = dentryManager_->GetDentry(parentId_, name_, &srcDentry_);
+    if (rc != CURVEFS_ERROR::OK) {
+        LOG(ERROR) << "Failed to get dentry, retCode = " << rc;
+        return rc;
+    }
+
+    rc = dentryManager_->GetDentry(newParentId_, newname_, &dstDentry_);
+    if (rc == CURVEFS_ERROR::NOTEXIST) {
+        return CURVEFS_ERROR::OK;
+    } else if (rc == CURVEFS_ERROR::OK) {
+        oldInodeId_ = dstDentry_.inodeid();
+        return CheckOverwrite();
+    }
+
+    LOG(ERROR) << "Failed to get dentry, retCode = " << rc;
+    return rc;
+}
+
+CURVEFS_ERROR RenameOperator::PrepareTx() {
+    dentry_ = Dentry(srcDentry_);
+    dentry_.set_txid(srcTxId_ + 1);
+    dentry_.set_flag(DentryFlag::DELETE_MARK_FLAG);
+
+    newDentry_ = Dentry(srcDentry_);
+    newDentry_.set_parentinodeid(newParentId_);
+    newDentry_.set_name(newname_);
+    newDentry_.set_txid(dstTxId_ + 1);
+
+    VLOG(1) << "Prepare rename tx: "
+            << "src dentry = (" << dentry_.ShortDebugString() << ")"
+            << ", dst dentry = (" << newDentry_.ShortDebugString() << ")";
+
+    auto rc = dentryManager_->Rename(dentry_, newDentry_);
+    if (rc != CURVEFS_ERROR::OK) {
+        LOG(ERROR) << "Rename failed, retCode = " << rc;
+    }
+    return rc;
+}
+
+CURVEFS_ERROR RenameOperator::CommitTx() {
+    std::vector<PartitionTxId> txIds;
+
+    PartitionTxId partitionTxId;
+    partitionTxId.set_partitionid(srcPartitionId_);
+    partitionTxId.set_txid(srcTxId_ + 1);
+    txIds.push_back(partitionTxId);
+
+    if (srcPartitionId_ != dstPartitionId_) {
+        partitionTxId.set_partitionid(dstPartitionId_);
+        partitionTxId.set_txid(dstTxId_ + 1);
+        txIds.push_back(partitionTxId);
+    }
+
+    auto succ = mdsClient_->CommitTx(txIds);
+    if (!succ) {
+        LOG(ERROR) << "CommitTx failed";
+        return CURVEFS_ERROR::INTERNAL;
+    }
+    return CURVEFS_ERROR::OK;
+}
+
+void RenameOperator::DeleteOldInode() {
+    if (oldInodeId_ == 0) {
+        return;
+    }
+
+    auto rc = inodeManager_->DeleteInode(oldInodeId_);
+    if (rc != CURVEFS_ERROR::OK) {
+        LOG(ERROR) << "DeleteInode fail, retCode = " << rc
+                   << ", inodeId = " << oldInodeId_;
+    }
+}
+
+void RenameOperator::UpdateCache() {
+    dentryManager_->DeleteCache(parentId_, name_);
+    dentryManager_->InsertOrReplaceCache(newDentry_, true);
```

Comment:
This insert may cause cache in dentry not consist with metaserver, maybe should get dentry from metaserver instead.

Reply:
> This insert may cause cache in dentry not consist with metaserver, maybe should get dentry from metaserver instead.

I think it will not cause the consistent problem, because the update operator after commit tx.